internal/flow/internal/controller/loader.go

```diff
@@ -539,6 +554,21 @@ func (l *Loader) populateComponentNodes(g *dag.Graph, componentBlocks []*ast.Blo
 	return diags
 }
 
+// blockAlreadyDefined returns (diag, true) if the given id is already in the provided blockMap.
+// else it adds the block to the map and returns (empty diag, false).
+func blockAlreadyDefined(blockMap map[string]*ast.BlockStmt, id string, block *ast.BlockStmt) (diag.Diagnostic, bool) {
+	if orig, redefined := blockMap[id]; redefined {
+		return diag.Diagnostic{
+			Severity: diag.SeverityLevelError,
+			Message:  fmt.Sprintf("block %s already declared at %s", id, ast.StartPos(orig).Position()),
+			StartPos: block.NamePos.Position(),
+			EndPos:   block.NamePos.Add(len(id) - 1).Position(),
+		}, true
+	}
+	blockMap[id] = block
+	return diag.Diagnostic{}, false
+}
```

Comment:
nit: I have to scroll pretty far down from where this is first used to where it is defined, which makes it difficult to follow along with the logic. Can you move this function to be directly under its first usage?